# Import libraries

In [1]:
import keras
from keras.models import Sequential
from keras.layers import Dense
from keras.utils import to_categorical
from keras.layers.convolutional import Conv2D # to add convolutional layers
from keras.layers.convolutional import MaxPooling2D # to add pooling layers
from keras.layers import Flatten # to flatten data for fully connected layers
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
import numpy as np
import pandas as pd

C:\Users\Hamad\anaconda3\lib\site-packages\scipy\__init__.py:146: UserWarning: A NumPy version >=1.16.5 and <1.23.0 is required for this version of SciPy (detected version 1.24.2
  warnings.warn(f"A NumPy version >={np_minversion} and <{np_maxversion}"


# Build the neural network

- One hidden layer of 10 nodes, and a ReLU activation function

- Use the adam optimizer and the mean squared error  as the loss function.

In [2]:
def NN():
    
    model = Sequential()
    
    model.add(Dense(units = 10, activation = 'relu'))
    model.add(Dense(units = 1, activation = 'linear'))
    
    model.compile(optimizer='adam', loss='mean_squared_error')
    
    return model

# Read the data

In [3]:
df = pd.read_csv("concrete_data.csv")
df.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

# Split the data into X and Y

In [4]:
X = df.drop('Strength', axis = 1)
y = df['Strength']

print(X.shape)
print(y.shape)

(1030, 8)
(1030,)


# Normalize the data

- Use a normalized version of the data. Recall that one way to normalize the data is by subtracting the mean from the individual predictors and dividing by the standard deviation.

In [5]:
X_norm = (X - X.mean()) / X.std()
X_norm.describe()

Cement  Blast Furnace Slag       Fly Ash         Water  \
count  1.030000e+03        1.030000e+03  1.030000e+03  1.030000e+03   
mean   2.432224e-15       -8.513686e-16  3.837815e-16  1.846743e-15   
std    1.000000e+00        1.000000e+00  1.000000e+00  1.000000e+00   
min   -1.714421e+00       -8.564718e-01 -8.467326e-01 -2.798851e+00   
25%   -8.496407e-01       -8.564718e-01 -8.467326e-01 -7.805147e-01   
50%   -7.911350e-02       -6.014861e-01 -8.467326e-01  1.607513e-01   
75%    6.586406e-01        8.003558e-01  1.001791e+00  4.885554e-01   
max    2.476712e+00        3.309068e+00  2.279976e+00  3.064159e+00   

       Superplasticizer  Coarse Aggregate  Fine Aggregate           Age  
count      1.030000e+03      1.030000e+03    1.030000e+03  1.030000e+03  
mean      -9.641155e-16      6.818710e-15    1.232571e-14  3.640022e-16  
std        1.000000e+00      1.000000e+00    1.000000e+00  1.000000e+00  
min       -1.038638e+00     -2.211064e+00   -2.239829e+00 -7.070160e-01  
25%       -1.038638e+00     -5.262618e-01   -5.317114e-01 -6.120340e-01  
50%        3.269920e-02     -6.326279e-02    7.383152e-02 -2.795973e-01  
75%        6.688058e-01      7.264077e-01    6.288606e-01  1.636517e-01  
max        4.351528e+00      2.213149e+00    2.731735e+00  5.055221e+00

# 1. Split the data into training and testing

- Randomly split the data into a training and test sets by holding 30% of the data for testing. You can use the train_test_splithelper function from Scikit-learn.

In [6]:
def data_split(X, y):
    return train_test_split(X, y, test_size=0.3)

X_train, X_test, y_train, y_test = data_split(X_norm, y)

# 2. Train the model

- Train the model on the training data using 50 epochs.

In [7]:
def model_training(X_train, y_train, verbose = 0):
    model = NN()
    model.fit(X_train, y_train, epochs = 50, verbose = verbose)
    return model

model = model_training(X_train, y_train, 1)

Epoch 1/50
23/23 [==============================] - 1s 2ms/step - loss: 1619.2190
Epoch 2/50
23/23 [==============================] - 0s 1ms/step - loss: 1601.6141
Epoch 3/50
23/23 [==============================] - 0s 1ms/step - loss: 1584.7405
Epoch 4/50
23/23 [==============================] - 0s 1ms/step - loss: 1567.9562
Epoch 5/50
23/23 [==============================] - 0s 1ms/step - loss: 1551.5920
Epoch 6/50
23/23 [==============================] - 0s 1ms/step - loss: 1534.7792
Epoch 7/50
23/23 [==============================] - 0s 1ms/step - loss: 1518.1152
Epoch 8/50
23/23 [==============================] - 0s 1ms/step - loss: 1501.0817
Epoch 9/50
23/23 [==============================] - 0s 1ms/step - loss: 1483.3632
Epoch 10/50
23/23 [==============================] - 0s 1ms/step - loss: 1465.6947
Epoch 11/50
23/23 [==============================] - 0s 1ms/step - loss: 1447.0363
Epoch 12/50
23/23 [==============================] - 0s 1ms/step - loss: 1428.2997
Epoch 13/50
2

# 3. Evaluate the model

- Evaluate the model on the test data and compute the mean squared error between the predicted concrete strength and the actual concrete strength. You can use the mean_squared_error function from Scikit-learn.

In [8]:
def prediction(model, X_test, y_test):
    y_pred = model.predict(X_test)
    return mean_squared_error(y_test, y_pred)

print(f'MSE : {prediction(model, X_test, y_test)}')

10/10 [==============================] - 0s 853us/step
MSE : 699.4354518748102


# 4. Repeat steps 1 to 3

- Repeat steps 1 - 3, 50 times, i.e., create a list of 50 mean squared errors.

In [9]:
MSE_lst = []

for i in range (50):
    X_train, X_test, y_train, y_test = data_split(X_norm, y)
    print(f'Model {i+1} :-')
    model = model_training(X_train, y_train)
    mse = prediction(model, X_test, y_test)
    MSE_lst.append(mse)

Model 1 :-
10/10 [==============================] - 0s 1ms/step
Model 2 :-
10/10 [==============================] - 0s 975us/step
Model 3 :-
10/10 [==============================] - 0s 1ms/step
Model 4 :-
10/10 [==============================] - 0s 3ms/step
Model 5 :-
10/10 [==============================] - 0s 1ms/step
Model 6 :-
10/10 [==============================] - 0s 2ms/step
Model 7 :-
10/10 [==============================] - 0s 2ms/step
Model 8 :-
10/10 [==============================] - 0s 2ms/step
Model 9 :-
10/10 [==============================] - 0s 2ms/step
Model 10 :-
10/10 [==============================] - 0s 2ms/step
Model 11 :-
10/10 [==============================] - 0s 839us/step
Model 12 :-
10/10 [==============================] - 0s 2ms/step
Model 13 :-
10/10 [==============================] - 0s 849us/step
Model 14 :-
10/10 [==============================] - 0s 3ms/step
Model 15 :-
10/10 [==============================] - 0s 3ms/step
Model 16 :-
10/10 [=========

# Report for MSE

In [10]:
dic = {
    'MSE' : []
}

models = []

for i in range(len(MSE_lst)):
    models.append(i+1)
    dic['MSE'].append(MSE_lst[i])
    
report = pd.DataFrame(dic)
report.index = models
report

MSE
1   331.339828
2   327.278469
3   351.002894
4   535.452547
5   428.827595
6   459.989142
7   328.539104
8   304.507195
9   260.897069
10  400.617059
11  348.852933
12  286.114999
13  235.547177
14  371.749396
15  367.492097
16  363.487193
17  354.230816
18  344.528346
19  406.231721
20  272.171353
21  307.118117
22  346.086533
23  273.619696
24  403.826332
25  424.463350
26  264.952429
27  285.772254
28  315.386776
29  297.192782
30  246.829881
31  283.783824
32  320.082286
33  442.371954
34  347.135816
35  355.989465
36  330.240143
37  374.709709
38  433.656220
39  324.215768
40  372.314673
41  262.939928
42  271.986898
43  308.098353
44  348.386739
45  296.950062
46  752.222708
47  271.731297
48  319.505047
49  331.247887
50  450.535985

# Report the mean and the standard deviation of the mean squared errors.

In [11]:
print(f"Average of MSE         : {np.mean(MSE_lst)}")
print(f"Standard Deviation MSE : {np.std(MSE_lst)}")

Average of MSE         : 348.84419686712766
Standard Deviation MSE : 84.45335210710674


# How does the mean of the mean squared errors compare to that from Step A?

We can see the Standard Deviation in the MSE has reduced from 435.00 to 84.45.
That is because of normalization, which has made the model much more consistant than before.